File: HotelReservationServiceJava/Room.java

import java.util.Arrays;
import java.util.stream.Stream;

public class Room
{
    //keeps track of assigned ids
    private static int idCount = 1;
    
    int id;
    int capacity;
    //One slot for each day of the month
    Reservation[] availability = new Reservation[30];
    double perPersonPrice;
    
    public Room(){
        this.id = idCount++;
    }
    
    public Room(int capacity, double perPersonPrice){
        this();
        this.capacity = capacity;
        this.perPersonPrice = perPersonPrice;
    }
    
    public int getId(){
        return id;
    }
    
    public Reservation[] getReservations(){
        return availability;
    }
    
    //Adds reservation only if there are available slots for the duration
    public boolean addReservation(Reservation res){
        boolean result = true;
        
        for(int i = res.getArrival() - 1; i < res.getArrival() + res.getDuration() - 1; i++){
            result &= (availability[i] == null);
        }
        if(result){
            for(int i = res.getArrival(); i < res.getArrival() + res.getDuration() - 1; i++){
                availability[i] = res;
            }
        }
        return result;
    }
    
    //remove reservation from availability array
    public boolean cancelReservation(int resId){
        for(int i = 0; i < 30; i++){
            if(availability[i] != null && availability[i].getId() == resId){
                availability[i] = null;
            }
        }
        return true;
    }
    
    //Returns completeness percentage of availability array
    public int completenessPercentage(){
        return Arrays.stream(availability).filter(x -> x != null)
                        .mapToInt(x -> 1).sum() * 10 / 3;
    }
    
    
    //returns total cost of all reservations on availability array
    public double totalCost(){
        return Arrays.stream(availability).filter(x -> x != null)
                        .mapToDouble(x -> perPersonPrice * x.getPersonCount()).sum();
    }
    
}

File: HotelReservationServiceCpp/src/Room.cpp
#include "Room.h"
#include "Reservation.h"

int Room::idCount = 1;

Room::Room(int cap, double ppPrice) :
    capacity{cap},
    perPersonPrice{ppPrice}
    {
        this->id = idCount++;
    }

int Room::getId(){
    return this->id;
}


Reservation * (&Room::getReservations())[30]{
    return availability;
}

bool Room::addReservation(Reservation& reservation){
    bool result = true;

    for(int i = reservation.getArrival() - 1; i < reservation.getArrival() + reservation.getDuration() - 1; i++){
        result &= (availability[i] == nullptr);
    }
    if(result){
        for(int i = reservation.getArrival(); i < reservation.getArrival() + reservation.getDuration() - 1; i++){
            availability[i] = &reservation;
        }
    }
    return result;
}

bool Room::cancelReservation(int resId){
    for(int i = 0; i < 30; i++){
        if(availability[i] != nullptr && availability[i]->getId() == resId){
            availability[i] = nullptr;
        }
    }
    return true;
}

double Room::totalCost(){
    double cost = 0;
    for(int i = 0; i < 30; i++){
        if(availability[i] != nullptr){
            cost += perPersonPrice * availability[i]->getPersonCount();
        }
    }
    return cost;
}

int Room::completenessPercentage(){
    int nullEntries = 0;
    for(int i = 0; i < 30; i++){
        if(availability[i] == nullptr){
            nullEntries++;
        }
    }
    return (30 - nullEntries) * (10 / 3);
}

Room::~Room(){}
